diff --git a/modules/openxr/openxr_api.cpp b/modules/openxr/openxr_api.cpp
index 102af19..83e882c 100644
--- a/modules/openxr/openxr_api.cpp
+++ b/modules/openxr/openxr_api.cpp
@@ -2705,11 +2705,30 @@ void OpenXRAPI::end_frame() {
 		}
 	}
 
+	Vector<OrderedCompositionLayer> ordered_layers_list;
+	bool projection_layer_is_first = true;
+
+	for (OpenXRExtensionWrapper *extension : composition_layer_providers) {
+		for (int i = 0; i < extension->get_composition_layer_count(); i++) {
+			OrderedCompositionLayer layer = {
+				extension->get_composition_layer(i),
+				extension->get_composition_layer_order(i),
+			};
+			if (layer.composition_layer) {
+				ordered_layers_list.push_back(layer);
+				if (layer.sort_order == 0) {
+					WARN_PRINT_ONCE_ED("Composition layer returned sort order 0, it may be overwritten by projection layer.");
+				} else if (layer.sort_order < 0) {
+					projection_layer_is_first = false;
+				}
+			}
+		}
+	}
+
 	// must have:
 	// - should_render set to true
-	// - a valid view pose for projection_views[eye].pose to submit layer
-	// - an image to render
-	if (!render_state.should_render || !render_state.view_pose_valid || (render_state.submit_projection_layer && !render_state.main_swapchains[OPENXR_SWAPCHAIN_COLOR].is_image_acquired())) {
+	// - when submitting projection, a valid view pose and acquired color image
+	if (!render_state.should_render || (render_state.submit_projection_layer && (!render_state.view_pose_valid || !render_state.main_swapchains[OPENXR_SWAPCHAIN_COLOR].is_image_acquired()))) {
 		// submit 0 layers when we shouldn't render
 		XrFrameEndInfo frame_end_info = {
 			XR_TYPE_FRAME_END_INFO, // type
@@ -2745,27 +2764,6 @@ void OpenXRAPI::end_frame() {
 		}
 	}
 
-	Vector<OrderedCompositionLayer> ordered_layers_list;
-	bool projection_layer_is_first = true;
-
-	// Add composition layers from providers
-	for (OpenXRExtensionWrapper *extension : composition_layer_providers) {
-		for (int i = 0; i < extension->get_composition_layer_count(); i++) {
-			OrderedCompositionLayer layer = {
-				extension->get_composition_layer(i),
-				extension->get_composition_layer_order(i),
-			};
-			if (layer.composition_layer) {
-				ordered_layers_list.push_back(layer);
-				if (layer.sort_order == 0) {
-					WARN_PRINT_ONCE_ED("Composition layer returned sort order 0, it may be overwritten by projection layer.");
-				} else if (layer.sort_order < 0) {
-					projection_layer_is_first = false;
-				}
-			}
-		}
-	}
-
 	if (render_state.submit_projection_layer) {
 		XrCompositionLayerFlags layer_flags = XR_COMPOSITION_LAYER_CORRECT_CHROMATIC_ABERRATION_BIT;
 		if (!projection_layer_is_first || render_state.environment_blend_mode != XR_ENVIRONMENT_BLEND_MODE_OPAQUE) {
